src/main/scala/mesosphere/marathon/MarathonScheduler.scala

```diff
@@ -187,18 +188,31 @@ class MarathonScheduler @Inject() (
 
   override def error(driver: SchedulerDriver, message: String) {
     log.warn("Error: %s".format(message))
-    suicide()
+
+    // Currently, it's pretty hard to disambiguate this error from other causes of framework errors.
+    // Watch MESOS-2522 which will add a reason field for framework errors to help with this.
+    // For now the frameworkId is removed for all messages.
+    val removeFrameworkId = true
+    suicide(removeFrameworkId)
   }
 
-  private def suicide(): Unit = {
-    log.fatal("Committing suicide")
+  /**
+    * Exit the JVM process.
+    * Optionally deleting the frameworkId. Depending on the cause of this error,
+    * Marathon is marked completed and can never re-register until the leading
+    * Mesos master process is kicked over, or the persisted FrameworkID is manually deleted.
+    *
+    * @param removeFrameworkId if true delete the stored frameworkId.
+    */
+  private def suicide(removeFrameworkId: Boolean): Unit = {
+    log.fatal(s"Committing suicide!")
+
+    if (removeFrameworkId) Await.ready(frameworkIdUtil.expunge, config.zkTimeoutDuration)
 
-    //scalastyle:off magic.number
     // Asynchronously call sys.exit() to avoid deadlock due to the JVM shutdown hooks
-    Future {
-      sys.exit(9)
-    } onFailure {
-      case t: Throwable => log.fatal("Exception while committing suicide", t)
+    // scalastyle:off magic.number
```

Comment:
This turns off the rule for the remainder of the file.  Please use `//scalastyle:on magic.number` following the next line or move the scalastyle directive to the end of the line, e.g.:

```scala
Future(sys.exit(9)).onFailure {  // scalastyle:off magic.number
```